rclpy/src/rclpy/context.cpp

```diff
@@ -20,103 +20,137 @@
+          PyErr_WarnFormat(
+            PyExc_RuntimeWarning, stack_level,
+            "[rclpy| %s : %s ]: failed to fini rcl_context_t: %s",
+            RCUTILS_STRINGIFY(__FILE__), RCUTILS_STRINGIFY(__LINE__), rcl_get_error_string().str);
+          rcl_reset_error();
+        }
+      }
+      delete context;
+    });
+  *rcl_context_ = rcl_get_zero_initialized_context();
+
+  // turn the arguments into an array of C-style strings
+  std::vector<const char *> arg_c_values(pyargs.size());
+  for (size_t i = 0; i < pyargs.size(); ++i) {
+    // CPython owns const char * memory - no need to free it
+    arg_c_values[i] = PyUnicode_AsUTF8(pyargs[i].ptr());
+    if (!arg_c_values[i]) {
+      throw py::error_already_set();
+    }
   }
 
-  size_t domain_id;
-  rcl_ret_t ret = rcl_context_get_domain_id(context, &domain_id);
+  InitOptions init_options(rcl_get_default_allocator());
+
+  // Set domain id
+  rcl_ret_t ret = rcl_init_options_set_domain_id(&init_options.rcl_options, domain_id);
   if (RCL_RET_OK != ret) {
-    throw RCLError("Failed to get domain id");
+    throw RCLError("failed to set domain id to init options");
   }
 
-  return domain_id;
+  if (arg_c_values.size() > static_cast<size_t>(std::numeric_limits<int>::max())) {
+    throw std::range_error("Too many cli arguments");
+  }
+  int argc = static_cast<int>(arg_c_values.size());
+  const char ** argv = argc > 0 ? &(arg_c_values[0]) : nullptr;
+  ret = rcl_init(argc, argv, &init_options.rcl_options, rcl_context_.get());
+  if (RCL_RET_OK != ret) {
+    throw RCLError("failed to initialize rcl");
+  }
+
+  throw_if_unparsed_ros_args(pyargs, rcl_context_.get()->global_arguments);
 }
 
 void
-_rclpy_context_handle_destructor(void * p)
+Context::destroy()
 {
-  auto context = static_cast<rcl_context_t *>(p);
-  if (!context) {
-    // Warning should use line number of the current stack frame
-    int stack_level = 1;
-    PyErr_WarnFormat(
-      PyExc_RuntimeWarning, stack_level, "_rclpy_context_handle_destructor failed to get pointer");
-    return;
-  }
-  if (NULL != context->impl) {
-    rcl_ret_t ret;
-    if (rcl_context_is_valid(context)) {
-      // shutdown first, if still valid
-      ret = rcl_shutdown(context);
-      if (RCL_RET_OK != ret) {
-        fprintf(
-          stderr,
-          "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
-          "failed to shutdown rcl_context_t (%d) during PyCapsule destructor: %s\n",
-          ret,
-          rcl_get_error_string().str);
-        rcl_reset_error();
-      }
-    }
-    ret = rcl_context_fini(context);
-    if (RCL_RET_OK != ret) {
-      fprintf(
-        stderr,
-        "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
-        "failed to fini rcl_context_t (%d) during PyCapsule destructor: %s\n",
-        ret,
-        rcl_get_error_string().str);
-      rcl_reset_error();
-    }
-  }
-  PyMem_FREE(context);
+  rcl_context_.reset();
```

Comment:
@sloretz meta: one thing that's probably not great is that, if it weren't for the consistent `with obj:` where obj is a `Destroyable`, you could easily crash the process by calling APIs after an explicit `destroy()` -- even in a single threaded setting where there's no risk for concurrent destruction making that `with` block somewhat unnecessary.

Reply:
Hmm yeah, maybe that `with obj:` logic could be moved into the APIs themselves.  Maybe a `DestroyableScopeGuard` or something with https://pybind11.readthedocs.io/en/stable/advanced/functions.html#call-guard that did the equivalent of `__enter__` and `__exit__` could replace that, and get rid of one of the jumps between Python and C++ at the same time.

Reply:
An implicit call guard would do, yeah.